numpy/_core/src/multiarray/arraywrap.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Definitions for dealing with array-wrap or array-prepare.
+ */
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include <Python.h>
+
+#include "numpy/arrayobject.h"
+#include "numpy/npy_3kcompat.h"
+#include "get_attr_string.h"
+
+#include "arraywrap.h"
+#include "multiarraymodule.h"
+
+
+/*
+ * Find the array wrap or array prepare method that applies to the inputs.
+ * outputs should NOT be passed, as they are considered individually while
+ * applying the wrapping.
+ *
+ * @param nin number of inputs
+ * @param inputs Original input objects
+ * @param out_wrap Set to the python callable or None (on success).
+ * @param out_wrap_type Set to the type belonging to the wrapper.
+ */
+NPY_NO_EXPORT int
+npy_find_array_wrap(
+        int nin, PyObject *const *inputs,
+        PyObject **out_wrap, PyObject **out_wrap_type)
+{
+    PyObject *wrap = NULL;
+    PyObject *wrap_type = NULL;
+
+    double priority = 0;  /* silence uninitialized warning */
+
+    /*
+     * Iterate through all inputs taking the first one with an __array_wrap__
+     * and replace it if a later one has a higher priority.
+     * (Currently even priority=-inf can be picked if it is the only argument.)
+     */
+    for (int i = 0; i < nin; i++) {
+        PyObject *obj = inputs[i];
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < NPY_PRIORITY) {
```

Comment:
I'm still confused: if you initialize `priority = -inf` (or some other very large negative number), will this line not just always be true if `wrap` is undefined? So, why is the `wrap == NULL` needed?

Reply:
not if the object defines `priority = -inf`.  Maybe that would be preferable, but :shrug:.

Reply:
Ah, I only need it on the last branch, yes.  Would that seem clearer (then the last branch is still special, this way priority is only initialized to silence warnings)?

Reply:
I think coding against something that defines `priority = -inf` yet defined `__array_wrap__` seems unnecessary! But if you really prefer what you have, that is fine too -- my own sense is that the logic becomes clearer without the extra conditions that are needed to grab the first argument unconditionally.

Reply:
I don't disagree, I think I just wanted to match *one* of the implementations inside NumPy.
Maybe it is laziness speaking: This way no user can run into this (not even hypothesis) and its there already there...?